src/ArborX_HDBSCAN.hpp

```diff
@@ -0,0 +1,163 @@
+/****************************************************************************
+ * Copyright (c) 2017-2021 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#ifndef ARBORX_HDBSCAN_HPP
+#define ARBORX_HDBSCAN_HPP
+
+#include <ArborX_DetailsHDBSCAN.hpp>
+#include <ArborX_DetailsMutualReachabilityDistance.hpp>
+#include <ArborX_DetailsTreeNodeLabeling.hpp>
+#include <ArborX_LinearBVH.hpp>
+
+#include <iomanip>
+#include <limits>
+#include <map>
+
+namespace ArborX
+{
+
+namespace HDBSCAN
+{
+
+struct Parameters
+{
+  bool _verbose = false;
+
+  Parameters &setVerbose(bool verbose)
+  {
+    _verbose = verbose;
+    return *this;
+  }
+};
+
+} // namespace HDBSCAN
+
+template <typename ExecutionSpace, typename Primitives>
+void hdbscan(ExecutionSpace const &exec_space, Primitives const &primitives,
+             int core_min_size,
+             HDBSCAN::Parameters const &parameters = HDBSCAN::Parameters())
+{
+  using MemorySpace = typename Primitives::memory_space;
+
+  Kokkos::Profiling::pushRegion("ArborX::HDBSCAN");
+
+  // Right now, we use the same minpts for computing core distances as well as
+  // minimum cluster size. For the latter, minpts = 2 is special in that it
+  // requires introducing a self-loops at the MST level to compute cluster
```

Comment:
In the 2015 paper, the MST is extended by adding edges `(i, i, core_distance(i))`. When we thought with @piyush314 about it, we decided that for `minpts > 2` we can avoid adding them. It may be that we'll need to add them anyway.